clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -0,0 +1,294 @@
+            new ConsumerGroupHeartbeatRequest.Builder(data),
+            coordinatorRequestManager.coordinator());
+
+        request.future().whenComplete((response, exception) -> {
+            if (exception == null) {
+                onResponse((ConsumerGroupHeartbeatResponse) response.responseBody(), response.receivedTimeMs());
+            } else {
+                onFailure(exception, response.receivedTimeMs());
+            }
+        });
+        return request;
+    }
+
+    private void onFailure(final Throwable exception, final long responseTimeMs) {
+        this.heartbeatRequestState.onFailedAttempt(responseTimeMs);
+        logger.warn("Failed to send heartbeat to coordinator node {} due to error: {}",
+                coordinatorRequestManager.coordinator(), exception.getMessage());
+    }
+
+    private void onResponse(final ConsumerGroupHeartbeatResponse response, long currentTimeMs) {
+        if (Errors.forCode(response.data().errorCode()) == Errors.NONE) {
+            this.heartbeatRequestState.updateHeartbeatIntervalMs(response.data().heartbeatIntervalMs());
+            this.heartbeatRequestState.onSuccessfulAttempt(currentTimeMs);
+            this.heartbeatRequestState.resetTimer();
+            try {
+                membershipManager.updateState(response.data());
+            } catch (KafkaException e) {
+                logger.error("Received unexpected error in heartbeat response: {}", e.getMessage());
+            }
+            return;
+        }
+
+        onErrorResponse(response, currentTimeMs);
+    }
+
+    private void onErrorResponse(final ConsumerGroupHeartbeatResponse response,
+                                 final long currentTimeMs) {
+        this.heartbeatRequestState.onFailedAttempt(currentTimeMs);
+        Errors error = Errors.forCode(response.data().errorCode());
+        if (error == Errors.NOT_COORDINATOR || error == Errors.COORDINATOR_NOT_AVAILABLE) {
+            String errorMessage = String.format("Coordinator node {} is either not started or not valid. Retrying",
+                    coordinatorRequestManager);
+            logInfo(errorMessage, response, currentTimeMs);
+            coordinatorRequestManager.markCoordinatorUnknown(response.data().errorMessage(), currentTimeMs);
+        } else if (error == Errors.COORDINATOR_LOAD_IN_PROGRESS) {
+            // retry
+            String errorMessage = String.format("Heartbeat was not successful because the coordinator node {} is loading. Retrying",
+                    coordinatorRequestManager.coordinator());
+            logInfo(errorMessage, response, currentTimeMs);
+        } else {
+            onFatalErrorResponse(response);
+        }
+    }
+
+    private void onFatalErrorResponse(final ConsumerGroupHeartbeatResponse response) {
+        final Errors responseError = Errors.forCode(response.data().errorCode());
+        if (responseError == Errors.GROUP_AUTHORIZATION_FAILED) {
```

Comment:
This is a non-retriable exception, so I expect we should be calling `membershipManager.transitionToFailure()`? (same for all other fatal exceptions up to the `UNRELEASED_INSTANCE_ID`, which is properly doing the transition)

Reply:
the updated PR should invoke maybeTransitionToFailureState() on a few fatal exceptions.